MAINT: 1.16.0rc2 backports

tylerjereddy opened the pull request:
Backports included so far:

1. gh-23086
2. gh-23044
3. gh-23029
4. gh-23047
5. gh-23048
6. gh-23051
7. gh-23091
8. gh-23099
9. gh-23106
10. gh-23110
11. gh-23113


-------

TODO:

- [x] reactivate CI, perhaps test wheel builds again depending on how many backports we finally end up with
- [x] consider merging some of the open backport-labeled PRs/dealing with other recent issues; i.e., https://github.com/scipy/scipy/pulls?q=is%3Apr+is%3Aopen+label%3Abackport-candidate
- [x] check that everything is "ok" with us relative to NumPy in release cycle
- [x] Potential release notes adjustment to deal with based on an analysis of type hints/stubs project: https://github.com/scipy/scipy/pull/23098#issuecomment-2942130558
- [x] recheck release notes after above shims
- [x] carefully reread the full diff

rgommers committed:
DOC: add documentation for the new `use-system-libraries` build option

[docs only]

mgorny committed:
TST: update all `result_to_tuple=` callables to accept two arguments

Update all the `result_to_tuple=` callables to accept two arguments,
and remove the compatibility hack for one-argument callables, fixing
the test suite on PyPy3.11.  The hack did not account for
`inspect.signature(tuple)` including `*args, **keywords` on PyPy
which led to it being misdetected as not requiring the compatibility
wrapping.  Rather than add further complexity to the hack, let's just
update all the functions to accept two arguments.

In addition to running the test suite with the changes, I have also
tried replacing the original hack with an explicitly raised exception
to make sure that no `result_to_tuple=` are passed a single-parameter
function.

Fixes #23035.

mdhaber committed:
MAINT: ndimage.vectorized_filter: fix behavior of axes when length equals dimensionality of image

mdhaber committed:
MAINT: ndimage.vectorized_filter: adjust scalar size to match number of axes

abhishek-iitmadras committed:
Update windows.yml

abhishek-iitmadras committed:
Update wheels.yml

tylerjereddy committed:
BLD: use newer pythran

* It should be safe to increase the pythran
upper bound in `pyproject.toml` per:
https://github.com/scipy/scipy/pull/23013#discussion_r2096347143

[ci skip] [skip ci]

tylerjereddy commented on pyproject.toml:
```diff
@@ -27,9 +27,9 @@ requires = [
+    # The upper bound on pythran is pre-emptive only; 0.18.0
     # is released/working at the time of writing
-    "pythran>=0.14.0,<0.18.0",
+    "pythran>=0.14.0,<0.19.0",
```
See discussion at https://github.com/scipy/scipy/pull/23013#discussion_r2096347143.

With `pythran` `0.18.0` installed alongside this branch locally, the full testsuite seemed happy on x86-64 linux via `spin test -j 32`.

rgommers committed:
MAINT: fix SPDX license expressions for LAPACK, OpenBLAS, GCC

Closes gh-23061

[skip ci]

dschult committed:
BUG: add test for eq and ne with different shapes. correct code (#23091)

crusaderky committed:
TST: CI broken on pytest 8.4.0

DietBru committed:
BUG: signal.csd: Zero-pad for different size inputs (#23047)

* Restore zero-padding for different size inputs for `signal.csd`.

 ---------

Co-authored-by: Tyler Reddy <tyler.je.reddy@gmail.com>

mdhaber committed:
DOC: ndimage.vectorized_filter: update size/footprint documentation

ajz34 committed:
BUG: spatial.distance: yule implementation in `distance_metrics.h` (#23110)

* Fix the `yule` distance implementation in `distance_metrics.h` to match the
behavior prior to gh-17538, and also to match the pure Python implementation.

jorenham commented:
I'm not sure if this is the right place for it, but I noticed that there were a couple of changes in 1.16.0rc1 that the release notes did not mention: https://github.com/scipy/scipy-stubs/pulls?q=is%3Apr+is%3Aclosed+label%3A%22stubs%3A+api+change+-+undocumented%22

Some of those are just internal changes that are probably not worth mentioning. But removals such as https://github.com/scipy/scipy-stubs/pull/557 and https://github.com/scipy/scipy-stubs/pull/544 look pretty (release-)noteworthy to me.

tylerjereddy committed:
DOC: update 1.16.0 release notes

* Update the SciPy `1.16.0` release notes following
additional backport activity.

* Also add a missing entry to the release notes for:
https://github.com/scipy/scipy-stubs/pull/544

* Address several other missing entries in the release
notes based on an analysis of our type hints/stubs
per:
https://github.com/scipy/scipy/pull/23098#issuecomment-2942130558

tylerjereddy commented:
@dschult or @stefanv or @perimosocordiae -- can you comment on the public object removals Joren mentioned at https://github.com/scipy/scipy-stubs/pull/557? Is this something I can concisely describe in the release notes--how would you like it done?

perimosocordiae commented:
These look like the cleanup of many of the unintentionally-public helper functions that we deprecated when moving all the sparse formats to private modules (e.g., coo.py -> _coo.py).

Maybe we can recycle the release notes from when the deprecations originally happened?

tylerjereddy commented:
alright, I'm working on it locally

tylerjereddy commented:
Ok, I went through that list and made a few shims to the release notes locally. Some cases were borderline and I either pinged some folks for opinions in the relevant PRs or let myself off the hook a bit for changes that were simultaneously bug fixes and API shims (it may better to just mention these cases though, where there is a bug fix *and* the user gains a new keyword argument, so I think I'm slightly wrong on that perhaps).

Overall, that was very helpful Joren, so thanks. Even if folks decide against some of the relnotes additions, it is good for visibility to consider them and there is a very low chance we'd have caught them without the stubs work.

In a few cases where there were long lists of removed deprecated objects I simply linked to one of Joren's PRs or issues, which isn't great for permanence but is nice for concision.

tylerjereddy commented:
Current status is summarized in the original comment/TODO list above. We're now at parity with the merged backport PRs and I've decided to flush the full CI matrix (without wheels for now) to see where things stand there.

Some of the release notes additions might benefit from a bit of scrutiny (i.e., can you actually access `dfitpack`? that entry needed double backticks locally think).

I talked to Sebastian and Chuck this morning and I don't think there is much left to worry about re: NumPy breaks for us, but I'd probably spend the ~5 minutes to check their latest post-rc1 maintenance branch against this branch tomorrow before I perhaps test the wheel builds here as well (assuming no other blockers before that). This does mean we're slipping a little behind schedule, but this is getting closer I think and we don't actually want to beat NumPy out of the gate by too much anyway as a sort of excuse.

rgommers reviewed (commented):
Overall what's included now looks good to me, one suggestion for the release notes

rgommers committed:
DOC: improve docs for `-Duse-system-libraries` build option (#23113)

[docs only]

tylerjereddy commented:
fyi, I do see that the NumPy `2.3.0` stable release is now out, so I probably don't have many excuses left. I'll see what I can do to push forward this weekend.

tylerjereddy committed:
DOC: PR 23098 revisions

* Adjust the SciPy `1.16.0` release notes to reduce
the emphasis on removal of non-public functions per
feedback at:
https://github.com/scipy/scipy/pull/23098#discussion_r2128119840

tylerjereddy committed:
DOC: PR 23098 revisions [wheel build]

* Update the SciPy `1.16.0` release notes
following additional backport activity.

[wheel build]

tylerjereddy commented:
When I test locally on ARM Mac against NumPy `2.3.0` stable release on this branch, I don't see any problems (well, I do, but they just seem to be floating point issues I see with other NumPy versions as well, at the moment with `scipy/interpolate/tests/test_bary_rational.py::TestAAA::test_spiral_cleanup`). So, I think that means we're "ok" for `2.3.0` as far as I can tell.

We're back at parity with merged backports after I backported Ralf's system dep doc PR this evening, and I applied Ralf's suggested release note shim above.

I went ahead and pushed in a commit to attempt wheel builds--I'd really like to get RC2 out on Sunday if possible. If the CI is green, including wheels, I'll be tempted to move forward with that plan.

rgommers reviewed (approved):
Changes look good, CI seems happy and is almost done, plan to move ahead SGTM.

tylerjereddy commented:
Ok, I manually re-read the diff and release notes changes and everything matches with my expectations. Sounds like Ralf agrees and CI is passing, so let's get rolling for RC2.
